standalone-metastore/metastore-server/src/main/resources/package.jdo

```diff
@@ -1549,6 +1549,83 @@
         <column name="RM_DUMP_EXECUTION_ID"/>
       </index>
     </class>
+
+    <class name="MStoredProc" table="STORED_PROCS" identity-type="datastore" detachable="true">
+      <datastore-identity>
+        <column name="SP_ID"/>
+      </datastore-identity>
+      <field name="createTime">
+        <column name="CREATE_TIME" jdbc-type="integer" allows-null="false"/>
+      </field>
+      <field name="lastAccessTime">
+        <column name="LAST_ACCESS_TIME" jdbc-type="integer" allows-null="false"/>
+      </field>
+      <field name="database">
+        <column name="DB_ID" allows-null="false"/>
+      </field>
+      <field name="name">
+        <column name="NAME" length="256" jdbc-type="VARCHAR" allows-null="false"/>
+      </field>
+      <field name="owner">
+        <column name="OWNER_NAME" length="128" jdbc-type="VARCHAR" allows-null="false"/>
+      </field>
+      <field name="source">
+        <column name="SOURCE" jdbc-type="MEDIUMTEXT" allows-null="false"/>
+      </field>
+      <field name="language">
+        <column name="LANG" jdbc-type="varchar" length="128" allows-null="false"/>
+      </field>
+      <field name="returnType">
+        <column name="RET_TYPE" jdbc-type="varchar" length="128" allows-null="true"/>
+      </field>
+      <field name="parameters" table="SP_POS_ARGS" >
+        <collection element-type="MPosParam"/>
+        <join>
+            <column name="SP_ID"/>
+        </join>
+        <element>
+          <embedded>
+            <field name="name">
+              <column name="NAME" jdbc-type="varchar" length="256" allows-null="false"/>
+            </field>
+            <field name="type">
+              <column name="TYPE" jdbc-type="varchar" length="128" allows-null="false"/>
+            </field>
+            <field name="length">
+              <column name="length" jdbc-type="integer" allows-null="true"/>
+            </field>
+            <field name="scale">
+              <column name="scale" jdbc-type="integer" allows-null="true"/>
+            </field>
+            <field name="isOut">
+              <column name="OUT" allows-null="false"/>
+            </field>
+          </embedded>
+        </element>
+      </field>
```

Comment:
I think instead of storing the return_type/argument types and such in the metastore - as they would never participate in a query or anything "usefull"; they will just travel as payload in the messages.
Given the fact that they are effectively implicit data which can be figured out from the function defintion - I think we may leave it to the execution engine; it should be able to figure it out (since it should be able to use it) .

optionally; to give ourselfs(and users) some type of clarity we could add a "signature" string to the table - which could provide a human readable signature

Reply:
> I'm not sure if they never participate in a query. If one wants to discover the stored procedures which are currently stored in a DB and find out on what data they operate they would need to do some clumsy string manipulations on the signature.

I believe you are thinking about `information_schema` stuff - its not set in stone that we have to get all that data from the metastore db - for this case we might add a few UDFs parameter info into an array or something ; so we will still store simple things in the metastore - but we could transform it into more readable in. 

> Considering that other DB engines also store these information separately I would like to keep it as it is for now and see how it works in practice. Later on when we have multi language support we can revisit this issue.

yes it might be..but it would be better to revisit stuff like this if its really needed; and not after we have introduced "something" which later we should care for even if we don't want to

I still think there will be no real benefit of "storing it" in a decomposed manner - it will be harder to go forward in case stuff changes - and right now will not use it for anything ; so let's remove it..and add it only if there is a real need for it.

Reply:
Thanks @zeroflag for working on this , this is very cool. 

There are a few challenges with keeping redundant information outside of the procedure text that I can think of. One of them is that while the semantics of the procedure definition may be well defined, the representation for the other fields may not be defined clearly. Another usual challenge is that if there are any changes in the future, you will have to ensure backwards compatibility for those fields too.

Going through the specific implementation of type handling, it seems you are keeping length and scale in different fields. This is not done when we store types in HMS for Hive. Any reason for that? Also, checking the documentation, it seems HPL/SQL can apply some transformations to the field type. Are those transformations applied before storing the definition or later on?

Based on that, I think keeping a lean representation in HMS has its advantages like @kgyrtkirk mentioned, specifically if those fields are not actually being used for the time being. It's preferable to add those fields later on if needed, rather than backtracking and removing fields.

Reply:
Hey @jcamachor thanks for the feedback, I'm glad you chimed in. Maybe there is a misunderstanding, this information is not redundant in any way, we're making use of it when invoking a procedure. This information must be stored somewhere in some form the only question is what representation to use. Just to clarify, currently I store the signature in a structured way + procedure body (without the signature) in text. 

What was proposed (but @kgyrtkirk correct me if I'm wrong) as an alternative solution is store the signature together with the body and optionally add a signature string to the table as well (this would be redundant). I think regardless of the representation, backward compatibility will always be a concern.

One thing I don't like about storing the signature in text is that the assumption that invoking the procedure will always require parsing. This is only true for now because HPL/SQL is an AST interpreter but if we ever want to make this performant probably we'll need to introduce a byte code VM at some point in the near future. While creating a procedure needs parsing but invoking it wouldn't, if we stored the byte code. How would this work in this case? I suppose we can add runtime information into the byte code, but that's not always an option. For example postgres allows you to invoke procedures implemented in C where runtime information about the signature is not available. This might be one reason why they also choose to store the signature separately in a structured way.

Multi language support was already raised by customers and adding it would be the easiest if we had common bytecode for different languages. One might want to call a procedure implemented in language A from a different language B. Then A would need to use the parser of language B to get the signature information, if the signature was stored in text.

We can keep speculating on this, but at this point this is still an experimental and undocumented feature, I'm open to change it later if we have proof that one way is better than the other. 

But if we decide to go with the alternative solution from now on, I suggest we choose a language agnostic representation (JSON or whatever) of the signature, instead of the unparsed text.

> Also, checking the documentation, it seems HPL/SQL can apply some transformations to the field type. Are those transformations applied before storing the definition or later on?

Yes, and that only affects create table statements.

Reply:
> What was proposed (but @kgyrtkirk correct me if I'm wrong) as an alternative solution is store the signature together with the body and optionally add a signature string to the table as well (this would be redundant). I think regardless of the representation, backward compatibility will always be a concern.

That was a suggestion to provide a way to store a human readable 1 liner for the function - I never thinked that we should parse it; the system should rely on the definition - the human readable field would be just for "show" to be used sysdb related tables/etc.

> This is only true for now because HPL/SQL is an AST interpreter but if we ever want to make this performant probably we'll need to introduce a byte code VM at some point in the near future.

If at some point in time the "parsing" will prove to be a bottle neck - then we should address it at that point ; rather than try to address a problem which we may not even face - correct me if I'm wrong but I think we will only parse the function 1 time during the HS2 life cycle - which doesn't seem to be a big deal to me.

> One thing I don't like about storing the signature in text is that the assumption that invoking the procedure will always require parsing. This is only true for now because HPL/SQL is an AST interpreter but if we ever want to make this performant probably we'll need to introduce a byte code VM at some point in the near future. 

I think we should clarify/separate 2 things - for stored procedures usually there will be 2 languages at play:
* the "host" language - in this case HiveQL - which will accept the function definition
* the "foreign" language - which may or may not need parsing; in case of "HPL/SQL" it will but that will not be true for all languages

because this patch doesn't contain any HiveQL parser changes - I don't know how that part will work; could you give some example how these functions could be used by the user?
```
create procedure x(a integer) as $$
insert into t values(a);
$$ langugage hplsql;

call x(12);
```


> For example postgres allows you to invoke procedures implemented in C where runtime information about the signature is not available.

I don't know what feature you are refering to ; but afaik to register a function in postgres you should [run a `create function` stmt](https://www.postgresql.org/docs/9.1/sql-createfunction.html) - which will contain all the arguments/etc; and because of that I don't see any issue with other languages - as a create procedure call for any language; must contain the full defintion - which includes argument types/etc - so we will store that for any supported language..

Reply:
> That was a suggestion to provide a way to store a human readable..

I understand it is only for showing the user, but it it feels like an afterthought, I don't really like it. It adds redundancy and it only has a single use. The structured information can be used to generate this output, (or in fact other outputs, like showing the parameters in a table) or to supply meta programming information to various tools, like code completion tools or method finders.

For example, search for the return type:

```
int i = fn.. // find every function that starts with "fn" and returns an int
```

Although this is a search for the return type not on the parameter but it's the same problem since the return type is also part of the signature.

I we had higher order functions (in fact JavaScript has and that might be the 2nd supported language) then:

```
filter(fn.., listOfString); // find every function that starts with "fn" and takes a single string parameter
```
Method finder:

This might look like a bit scifi, there are only 2 programming environments I'm aware of, which know how to do this.

```
MethodFinder.methodFor('somefile.txt', 'txt'); // find which method returns the extension part of filename, by an by example,
```

It will return:
```
"FilenameUtils.getExtension()"
```

These things might sound unimportant but I think language design and tool development shouldn't be separated. Tool support should be considered from day 1 when creating a language.


> If at some point in time the "parsing" will prove to be a bottle neck

I'm was not talking about parsing as a bottleneck (though it could be) but interpreting the AST (method body) as a bottleneck. I can't think of any imperative language that can be taken seriously that works that way. Perhaps shell scripts, or early versions of Ruby which was notoriously slow so later they changed it.

> I think we should clarify/separate 2 things

If we only allow defining procedures in terms of the "host language", then this is true indeed. My assumption was that we might want to accept procedure definitions in terms of the "foreign language". For example function(x,y) {}. But ok, let's say this is not allowed. Then you're right, using the alternative format seems to be viable if we don't count the other issues.

If we go to that route, what would you do with the other columns which are also part of the signature, like "LANG", "RET_TYPE", "NAME" ?

```
CREATE FUNCTION func1(a int) RETURNS int LANGUAGE XX BEGIN ... END;
```

> I don't know what feature you are refering 

It's probably that one. I think that illustrates nicely why parsing is not always a prerequisite of calling a procedure.

Reply:
> I understand it is only for showing the user, but it it feels like an afterthought, I don't really like it.

Okay; then we don't need to add and store it - we could also use a simple UDF to get the signature if we want that.

About  `search for the return type` I think we should come up with a "working solution first" and go after these things when we need it (maybe never?)
I think that `MethodFinder.methodFor` will need a lot more than just a signature...so it's a little bit too much

> These things might sound unimportant but I think language design and tool development shouldn't be separated. Tool support should be considered from day 1 when creating a language.

We may leave this thing out completely right now - because we don't need it right now.
I think instead of adding something which may not perfectly align with our future needs ; leaving something out will not lock us in at all.

> but interpreting the AST (method body) as a bottleneck.

that's the problem of the language implementation itself - I think in case the language function is defined in some kind of "text" then we should store it as "text" - to have a less convoluted contract with the embedded language.

> If we only allow defining procedures in terms of the "host language", then this is true indeed. My assumption was that we might want to accept procedure definitions in terms of the "foreign language".

yes; we should only store those procedures/functions which are usable from the HiveQL - don't we?

> `CREATE FUNCTION func1(a int) RETURNS int LANGUAGE XX BEGIN ... END;`

I think we should simply store the whole definition from `CREATE` to the closing `;`.
Storing anything less could become unparsable.

> > I don't know what feature you are refering

> It's probably that one. I think that illustrates nicely why parsing is not always a prerequisite of calling a procedure.

There might be some misunderstanding here by "parsing" - I mean to process it with the host sql language. 

If the client language has performance issues doing parsing/etc is then that issue belong to the language itself.
In case the language is an on-the fly interpreted language which also happens to have a compiled version ; then that other version could either be registered as a separate language (and refer to a binary somehow) or the language could add some caching/etc to avoid unneccessary parsing overhead.
Couldn't we do something like this for hplsql?

In any case: I think that extending the basic implementation with a "generic blob storage" option to provide additional services for the stored procedures (which could potentially be used to speed up sql function execution) should be a separate feature - and as such; should be discussed(and implemented) separetly.
(Honestly I think there would be marginal benefits implementing this - and could be dodged by the client language implementation with a few caches/etc.)

Reply:
Discussed this further offline, let's try the string based approach for now and see how it goes. I'll modify the patch.